commit 9cb5518b7ff69ab3bbc422313c69d54696fe8472
Author: Tom Lane <tgl@sss.pgh.pa.us>
Date:   Tue Jan 11 12:03:06 2022 -0500

    Clean up error message reported after \password encryption failure.
    
    Experimenting with FIPS mode enabled, I saw
    
    regression=# \password joe
    Enter new password for user "joe":
    Enter it again:
    could not encrypt password: disabled for FIPS
    out of memory
    
    because PQencryptPasswordConn was still of the opinion that "out of
    memory" is always appropriate to print.
    
    Minor oversight in b69aba745.  Like that one, back-patch to v14.

diff --git a/src/interfaces/libpq/fe-auth.c b/src/interfaces/libpq/fe-auth.c
index 2e6b2e8f04e..82fc7cdb986 100644
--- a/src/interfaces/libpq/fe-auth.c
+++ b/src/interfaces/libpq/fe-auth.c
@@ -1288,10 +1288,14 @@ PQencryptPasswordConn(PGconn *conn, const char *passwd, const char *user,
 	 * Ok, now we know what algorithm to use
 	 */
 	if (strcmp(algorithm, "scram-sha-256") == 0)
 	{
 		crypt_pwd = pg_fe_scram_build_secret(passwd);
+		/* We assume the only possible failure is OOM */
+		if (!crypt_pwd)
+			appendPQExpBufferStr(&conn->errorMessage,
+								 libpq_gettext("out of memory\n"));
 	}
 	else if (strcmp(algorithm, "md5") == 0)
 	{
 		crypt_pwd = malloc(MD5_PASSWD_LEN + 1);
 		if (crypt_pwd)
@@ -1305,20 +1309,19 @@ PQencryptPasswordConn(PGconn *conn, const char *passwd, const char *user,
 								  errstr);
 				free(crypt_pwd);
 				crypt_pwd = NULL;
 			}
 		}
+		else
+			appendPQExpBufferStr(&conn->errorMessage,
+								 libpq_gettext("out of memory\n"));
 	}
 	else
 	{
 		appendPQExpBuffer(&conn->errorMessage,
 						  libpq_gettext("unrecognized password encryption algorithm \"%s\"\n"),
 						  algorithm);
 		return NULL;
 	}
 
-	if (!crypt_pwd)
-		appendPQExpBufferStr(&conn->errorMessage,
-							 libpq_gettext("out of memory\n"));
-
 	return crypt_pwd;
 }
